drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -250,43 +388,82 @@ const PORT_MAP: [PortLocation; 32] = [
+/// - The module interaction succeeded
+/// - The module interaction failed
+/// - The module could not be interacted with due to an FPGA communication error
+#[derive(Copy, Clone, Default, PartialEq)]
+pub struct ModuleResult {
+    success: LogicalPortMask,
+    failure: LogicalPortMask,
+    error: LogicalPortMask,
 }
 
-// Maps logical port `mask` to physical FPGA locations
-impl From<LogicalPortMask> for FpgaPortMasks {
-    fn from(mask: LogicalPortMask) -> FpgaPortMasks {
-        let mut fpga_port_masks = FpgaPortMasks::default();
-        for (i, port_loc) in PORT_MAP.iter().enumerate() {
-            let port_mask: u32 = 1 << i;
-            if (mask.0 & port_mask) != 0 {
-                match port_loc.controller {
-                    FpgaController::Left => {
-                        fpga_port_masks.left.set(port_loc.port);
-                    }
-                    FpgaController::Right => {
-                        fpga_port_masks.right.set(port_loc.port);
-                    }
-                }
-            }
+impl ModuleResult {
+    /// Create a new ModuleResult which enforces no overlap in the success,
+    /// failure, and error masks.
+    pub fn new(
+        success: LogicalPortMask,
+        failure: LogicalPortMask,
+        error: LogicalPortMask,
+    ) -> Result<Self, TransceiversError> {
+        if (success.0 & failure.0 != 0)
+            || (success.0 & error.0 != 0)
+            || (failure.0 & error.0 != 0)
+        {
+            return Err(TransceiversError::InvalidModuleResult);
         }
-        fpga_port_masks
+        Ok(Self {
+            success,
+            failure,
+            error,
+        })
     }
-}
 
-impl From<LogicalPort> for PortLocation {
-    fn from(port: LogicalPort) -> PortLocation {
-        PORT_MAP[port.0 as usize]
+    pub fn success(&self) -> LogicalPortMask {
+        self.success
+    }
+
+    pub fn failure(&self) -> LogicalPortMask {
+        self.failure
+    }
+
+    pub fn error(&self) -> LogicalPortMask {
+        self.error
+    }
+
+    /// This is a helper function to combine two sets of results for when there
+    /// is a sequence of `ModuleResult` yielding function calls. Building such a
+    /// sequence is generally done with the following form (where `modules` is
+    /// a `LogicalPortMask` of requested modules):
+    ///
+    /// let result = some_result_fn(modules);
+    /// let next_result = result.chain(another_result_fn(result.success))
+    ///
+    /// So the initial result includes some set of success, failure, and error
+    /// masks which then need to be reconciled with a new set of masks, generally
+    /// a subset of the success mask of the initial result. Notably, there
+    /// cannot be overlap between these masks, which this function enforces.
+    pub fn chain(&self, next: ModuleResult) -> Self {
+        // success mask is just what the success of the next step is
+        let success = next.success;
```

Comment:
Yeah, that is a miss on my part! I'll add this line above the current 458:
```rust
assert!(next.success().to_indices().all(|idx| self.success().is_set(LogicalPort(idx))));
```
This will assert that every index of found in `next.success` can also be found in `self.success`.